include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,485 @@
+    //!\hideinitializer
+    using size_type = typename ranges::v3::size_type_t<inner_type>;
+    //!\brief Use the difference_type of the underlying sequence.
+    //!\hideinitializer
+    using difference_type = typename ranges::v3::difference_type_t<inner_type>;
+
+
+    //!\brief The iterator type of this container (a random access iterator).
+    //!\hideinitializer
+    using iterator = detail::random_access_iterator<gap_decorator_anchor_set>;
+    //!\hideinitializer
+    using const_iterator = detail::random_access_iterator<gap_decorator_anchor_set const>;
+    //!\}
+
+    /*!\name Constructors/Destructors
+     * \{
+     */
+    //!\brief Default constructor.
+    constexpr gap_decorator_anchor_set() = default;
+    //!\brief Copy constructor.
+    constexpr gap_decorator_anchor_set(gap_decorator_anchor_set const &) = default;
+    //!\brief Copy construction via assignment.
+    constexpr gap_decorator_anchor_set & operator=(gap_decorator_anchor_set const &) = default;
+    //!\brief Move constructor.
+    constexpr gap_decorator_anchor_set (gap_decorator_anchor_set && rhs) = default;
+    //!\brief Move assignment.
+    constexpr gap_decorator_anchor_set & operator=(gap_decorator_anchor_set && rhs) = default;
+    //!\brief Direct sequence assignment resets previously inserted gaps.
+    constexpr gap_decorator_anchor_set & operator=(inner_type sequence)
+    {
+        this->sequence = &sequence;
+        anchors.clear();
+    };
+    //!\brief Use default deconstructor.
+    ~gap_decorator_anchor_set() = default;
+    //!\brief Construct by host and explicit position.
+    constexpr gap_decorator_anchor_set(inner_type const & sequence): sequence{&sequence} {};
+    //!\}
+
+    /*!\brief Returns the total length of the aligned sequence.
+     * \returns The total lengths of the aligned sequence.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * No-throw guarantee.
+     */
+    size_type size() const noexcept
+    {
+        if (anchors.size())
+            return (*(anchors.rbegin())).second + sequence->size();
+        return sequence->size();
+    }
+
+    /*!\brief Insert a gap of length size at the aligned sequence iterator position.
+    * \returns A boolean flag indicating the success of the insertion operation.
+    *
+    * \par Complexity
+    *
+    * Average and worst case (insertion before last gap): o(k),
+    * Best case (back insertion): o(log k).
+    *
+    * \par Exceptions
+    *
+    * May throw std::set exceptions.
+    */
+    bool insert_gap(iterator const it, size_type const size=1)
+    {
+        size_type const pos = it - this->begin();
+        if (pos > size())
+            return false;
+        anchor_set_iterator it_set = anchors.begin();
+        // case 1: extend previous/surrounding gap already existing
+        if ((pos < this->size()) && (((value_type)(*this)[pos] == gap::GAP) || (pos > 0 && (value_type)(*this)[pos-1] == gap::GAP)))
```

Comment:
Not exactly, I returned (now assert, because iterator and not bool flag is returned) when `pos > size()`, but `pos = size()` is a valid insertion position for gap (tailing gaps). With this additional if condition I omit searching in the anchor set, because in that case it is just an appending of a new gap handled in the `else` part.